rpc/handlers.go

```diff
@@ -1379,6 +1395,82 @@ func (h *Handler) SpecVersion() (string, *jsonrpc.Error) {
 	return "v0.5.0", nil
 }
 
+func randomID() uint64 {
+	var n uint64
+	for err := binary.Read(rand.Reader, binary.LittleEndian, &n); err != nil; {
+	}
+	return n
+}
+
+func (h *Handler) SubscribeNewHeads(ctx context.Context) (uint64, *jsonrpc.Error) {
+	w, ok := jsonrpc.ConnFromContext(ctx)
+	if !ok {
+		return 0, jsonrpc.Err(jsonrpc.MethodNotFound, nil)
+	}
+
+	id := randomID()
+	subscriptionCtx, subscriptionCtxCancel := context.WithCancel(ctx)
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.subscriptions[id] = &subscription{
+		cancel: subscriptionCtxCancel,
+		conn:   w,
+	}
+
+	go func() {
+		sink := make(chan *core.Header)
+		sub := h.bcReader.SubscribeNewHeads(sink)
+		defer func() {
+			sub.Unsubscribe() // Closes sink.
+			h.mu.Lock()
+			defer h.mu.Unlock()
+			delete(h.subscriptions, id)
+		}()
+		for {
+			select {
+			case <-subscriptionCtx.Done():
+				return
+			case err := <-sub.Err():
+				h.log.Warnw("Subscription failed", "err", err)
+				return
+			case header := <-sink:
+				resp, err := json.Marshal(jsonrpc.Request{
```

Comment:
why `Request` and not a `jsonrpc.response` ?

Reply:
Ethereum abuses the jsonrpc spec. I think the reasoning is that standard JSON-RPC 2.0 requests can be sent to the client without expecting a response; they just happen to contain information that is useful to the client.

I'm not sure how useful that is; I feel any reasonable client would need to implement support for subscriptions anyway.